File: backend/test/c/test_exceptions.c
#include <assert.h>
#include <string.h>

#include "test_exceptions_c.h"

int main()
{
  assert(test_add_nothrow(1, 2) == 3);

  bind_error_reset();
  test_add_throw_runtime(1, 2);
  assert(bind_error_what() && strcmp(bind_error_what(), "runtime error") == 0);

  bind_error_reset();
  test_add_throw_bogus(1, 2);
  assert(bind_error_what() && strcmp(bind_error_what(), "exception") == 0);

  return 0;
}

File: backend/test/rust/test_exceptions.rs
use std::fmt;
use std::ffi::*;
use std::os::raw::*;

include!("rust_bind_error.rs");

include!("test_exceptions_rust.rs");

fn main() {
    unsafe {

    assert!(test_add_nothrow(1, 2) == 3);

    let err = match test_add_throw_runtime(1, 2) {
        Ok(_)  => String::from("no exception"),
        Err(e) => e.what,
    };

    assert!(err == "runtime error");

    let err = match test_add_throw_bogus(1, 2) {
        Ok(_)  => String::from("no exception"),
        Err(e) => e.what,
    };

    assert!(err == "exception");

    }
}
